docs/sources.md

```diff
@@ -0,0 +1,268 @@
+So for a container image source, it would basically amount to the image digest (plus some other metadata).
+
+The purpose of `CacheKey()` is for the solver to be able to ask the source, "I have this `id`, what unique cache key is there for it?"
+If the response is identical to the one already in cache, the solver knows it does not need to retrieve the data again, and simply can
+use what is in its cache. If it is different or does not exist, then it knows it needs to ask the source for the content referred to by the
+`id`
+
+This is expected to be idempotent, a read-only activity that returns a unique cache key.
+
+The arguments to `CacheKey()` are:
+
+* `ctx context.Context`: A golang [context](https://pkg.go.dev/context#Context).
+* `g session.Group`: The [session.Group](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/session#Group) contains parameters to identify the specific buildkit client calling this build. These may be useful for things like credentials.
+* `index int`: The `CacheKey()` can be called multiple times for the same source item, each indicating that it is a separate subitem. The `index` indicates which iteration of call this is. Generally, if the `index > 0` then it is complete.
+
+The expected return values are:
+
+* `cacheKey string`: the unique cache ID; this might be the combination of a hash with the source identifier. For example, `docker-image` uses the hash of the image config, while `local` combines the path to the directory and hash of a json that includes the session and search parameters.
+* `imgDigest string`: the digest of the root of the image. For example, `docker-image` uses the hash of the root manifest or index, while `local` returns the hash of the constructed json of the session and search parameters.
+* `cacheOpts solver.CacheOpts`: The [CacheOpts](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/solver#CacheOpts) contain metadata that can be passed from one iteration of `CacheKey()` to another, enabling context to be passed from call to call without the `Source` needing to store information. This is useful for items like progress streams and remote reference information.
+* `cacheDone bool`: Indicates whether or not caching is complete.
+* `err error`: error, if any.
+
+#### `Snapshot`
+
+`Snapshot()` is what the solver calls to retrieve the content for the `id`. Normally, this gets run if it turns out there is no cached result. It does the work to
+actually run the operation and create the underlying cache reference in the buildkit worker.
+
+Note that it is not absolutely necessary to retrieve the entire content when `Snapshot()` is called. It is possible to do a "lazy retrieval". For example,
+`docker-image` can use lazy retrieval, and defer pulling layers until they really are needed. However, everything that is needed to retrieve the content gets set up
+by `Snapshot()`.
+
+For example, `docker-image` pulls the image, while `local` ???
```

Comment:
I have no idea what it does, actually.

And when `docker-image` "pulls" the image, does it store it in some local cache? Or is it just passing the data to the solver? How does it handle various layers? It looks (from the code) like it merges them together to create a single filesystem view?